File: 0153-find-minimum-in-rotated-sorted-array/0153-find-minimum-in-rotated-sorted-array.java
class Solution {
    public int findMin(int[] nums) {
        int times = nums.length;
        int l = 0;
        int r = nums.length - 1;
        int min = nums[0];
        while(l <= r){
            int mid = l +(r-l)/2;
            if(nums[l] > nums[r]){
                l++;
                min = Math.min(nums[mid],min);
            }else if(nums[l] < nums[r]){
                min = Math.min(min, nums[l]);
                r = mid - 1;
            }else{
                min = Math.min(nums[mid],min);
                break;
            }
            
        }
        return min;
        
    }
}

File: 0153-find-minimum-in-rotated-sorted-array/0153-find-minimum-in-rotated-sorted-array.py
class Solution:
    def findMin(self, nums: List[int]) -> int:
        #if left index > right index it has been rotated
        #binary search to find the minimum
        left = 0
        right = len(nums)-1
        
        while left < right:
            mid = (left+right)//2
            if nums[mid] > nums[right]:
                #have to minimize the search space if it si bigger then it has to be the one on its left
                #if i leave without +1 it will infinitely loop at 2 elements
                left = mid +1
            else:
                right = mid
            
        
        return nums[left]
            